### ISSUE TITLE
Support syntactic infinity

### ISSUE THREAD

**Issue by masak** [masak/alma#517]
Macros eliminate calls to themselves through expansion. Here's an example of a _term_ which eliminates itself through (best-effort) dataflow analysis and escape analysis. If it can't eliminate itself, it fails at CHECK time with an error.

`inf` is a literal without a corresponding evaluation. Instead, it must _provably_ end up in comparisons (`<`, `<=`, `==`, `!=`, `>=`, `>`). It evaluates the comparison as if it's larger than any other value except itself. (Ditto `-Inf` and "smaller than any other value".) (If we ever get a `nan` value for floats (which will be an actual value, not just a syntactic term), it might get an overriding exception, failing all comparisons except `!=`.)

Funnily enough, this will give us the much-desired "allomorphic inf" of Perl 6; this `inf` can "have all of the types" in effect by not being a value; only its role in a comparison can have a (boolean) value. Thus it will work for numbers, strings, booleans, enum types &mdash; anything comparable.

Example:

```
func max(values: Iterable) {
    my max = -inf;
    for values -> v {
        if max < v {
            max = v;
        }
    }
    if max == -inf {
        throw new X.Value { message: "Cannot take max of empty Iterable" };
    }
    return max;
}
```

This translates to something roughly like this:

```
func max(values: Iterable) {
    my max;
    my minusInfInMax = true;
    for values -> v {
        if minusInfInMax || max < v {
            max = v;
            minusInfInMax = false;
        }
    }
    if minusInfInMax {
        throw new X.Value { message: "Cannot take max of empty Iterable" };
    }
    return max;
}
```

Except that `minusInfInMax` is hygienically hidden and can't collide with any actual variable name in scope.

We have syntactic junctions (#210), so why not syntactic infinities?

**Comment by masak** [comment]
I guess we can also allow `inf` and `-inf` to be stringified/printed, too. The main point is that they can never (not be proven not to) escape, in the "escape analysis" sense.

**Comment by masak** [comment]
Maybe it's that I've only recently read about [antimemes](http://www.scp-wiki.net/antimemetics-division-hub) for the first time, but this issue felt similar.

**Comment by masak** [comment]
Ranges.

I also feel it's reasonable to be able to do this:

```
for 1..inf -> n {
    say(n);
}
```

And this:

```
n = +prompt("n: ");

if n ~~ -inf..0 {
    say("negative");
}
elsif n ~~ 0..10 {
    say("low");
}
elsif n ~~ 11..inf {
    say("high");
}
```

Since ranges aren't aware of `inf` and `-inf` at all, it falls on the macros surrounding the latter to "hijack" the former somehow and "poison" ranges-with-infinities in them and treat them analogously to individual infinities.

I'm sure the implementors can hammer out the minute details. 😈

**Comment by masak** [comment]
I think "syntactic infinity" might end up being a "cute today" idea that doesn't quite last all the way to general usefulness. As [this talk](https://www.youtube.com/watch?v=3yvrs9S0RIw) points out, "users want values"... and "syntactic" here means "not first-class".

Maybe trying to do the same neatly through the type system, like with the [extended real number line](https://en.wikipedia.org/wiki/Extended_real_number_line) or the [extended complex plane](https://en.wikipedia.org/wiki/Riemann_sphere) &mdash; adding infinities from the outside, as it were. The subset of cases that are purely syntactical could then maybe be the subject of really cool optimizations... if that ends up on the right side of a cost/benefit analysis, I dunno.

Closing, not because it ain't a cool idea, but because there are probably more urgent ideas.

**Comment by masak** [comment]
> Maybe trying to do the same neatly through the type system, like with the [extended real number line](https://en.wikipedia.org/wiki/Extended_real_number_line) or the [extended complex plane](https://en.wikipedia.org/wiki/Riemann_sphere) — adding infinities from the outside, as it were.

I came back here now, a year later, to make essentially this point. I was writing some code that needed an infinity &mdash; and this being Java, I ended up doing `int INFINITY = Integer.MAX_VALUE;` &mdash; a poor-man's solution because it lays claim to `Integer.MAX_VALUE` as a kind of fake out-of-band value. Technically I should probably require that no actual in-band values take on this value as a precondition.

_Anyway_, I felt as I did this that a "true" way to do this would be to define a _type_ locally, very similar to `Optional`/`Maybe` in that it adds an extra point &mdash; well, two extra points, since we probably want negative infinity while we're at it &mdash; and then just extends the comparison operation in the natural way. For some reason, this strikes me as "nicer" than a purely syntactic approach. Maybe one reason is that by talking about it as types and values, we can better reason about how they flow in the code; the corresponding reasoning about a purely syntactic extension feels brittle, because the syntax is so static, but we still need to worry about values escaping. By extending the type, values could still escape if we wanted; it's when they don't that we can fearlessly optimize things locally.
